### ISSUE TITLE
AssemblyPath contains invalid charaters

### ISSUE THREAD

**Issue by jrjrguo** [nunit/nunit#2898]
When running test on UWP, we ran into an ArgumentException (Illegal characters in path) during BuildTestAssembly in DefaultTestAssemblyBuilder.cs. It seems to be complaining about assemblyPath, which is "\<unknown\>" instead of the real assembly path. Here are the related codes:
https://github.com/nunit/nunit/blob/e94c6d9cd4655c63c514ce354f7e8d451dc0118a/src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs#L83
https://github.com/nunit/nunit/blob/e94c6d9cd4655c63c514ce354f7e8d451dc0118a/src/NUnitFramework/framework/Internal/AssemblyHelper.cs#L56
This issue was introduced in NUnit 3.7. Prior to 3.7, it calls `AssemblyHelper.GetAssemblyName(assembly).FullName` to get the assemblyPath, which works fine. Now, it all uses `Assembly.GetAssemblyPath(assembly)`. According to [.NET API](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.module.fullyqualifiedname?view=netframework-4.7.2#System_Reflection_Module_FullyQualifiedName), `FullyQualifiedName` for the module might be "\<unknown\>" in certain situations. Switching back to GetAssemblyName simply fixes the problem here.

I may be wrong, but it seems to me that the only use of assemblyPath is to get the name of the assembly later. So why not just getting the name instead of the path here?

**Comment by jrjrguo** [comment]
@jnm2 Just saw your message.

Sure, I would like to make the changes. I will take a deeper look today.

**Comment by jrjrguo** [comment]
I think there are two options to address this issue:

Option 1: Preserve the TestAssembly class: TestAssembly currently takes in 'assemblyPath' and gets the name by calling `Path.GetFileName`. Although 'assemblyPath' and 'assemblyName' are both used throughout, it seems they all mean path by default. In this way, the only thing that needs to be changes is the <unknown> case. Like you said, there can be a quick check on the returned path, and pass the assembly name instead if the path contains invalid characters. The changes would be minimal.

Option 2: Change TestAssembly so that it takes names instead of paths. There are that many usage so the change is trivial, although there might be a lot of renaming. In this way, the caller needs to get the name from the assembly first rather than passing the assemblyPath directly. This feels more like what you suggested. Personally, I think it's clearer because TestAssembly only uses name essentially, but I feel like TestAssembly was designed to take in path for a reason.

@jnm2 Opinions?

**Comment by jnm2** [comment]
I like option 2 best. It's misleading to ask for an assembly path because we know there won't always be one.

**Comment by oznetmaster** [comment]
I am not at a computer, but I believe that in nunit-lite, the location of the test assembly is not defaulted, and can in fact be anywhere.  Using only the assembly name may lose the actual location of the assembly?

**Comment by jnm2** [comment]
@oznetmaster I'm not sure in what way the assembly location would be lost? At the point where we are making the change, we already have a permanent reference to the Assembly object. I couldn't find any code that relied on using the suite name as an assembly path.

**Comment by jrjrguo** [comment]
I think @jnm2 is right. Here it already has an reference to the Assembly, and the only use of the path/name is to pass it to TestAssembly so that it can then get the name from it.

When I was making the pull request, I realized that some of the assemblyName and assemblyPath are misused - no matter what the name is, it can be name or path most of the time. So I renamed those to be assemblyNameOrPath to avoid confusion. In this way, we can still preserve that path and load the test assembly from the correct location. @oznetmaster Is this your concern?
